### PR TITLE
feat(cli): integrate python with shared install and artifact lifecycles

### PR DESCRIPTION
**Pull request by zkochan**
## Summary

Replaces the fixture-only Python spike with an opt-in Python integration through the real pnpm v12 CLI, then extracts the shared install lifecycle exercised by npm, Cargo and Python. Related to pnpm/pnpm#14566 and pnpm/rfcs#34.

- `python.enabled` configuration, workspace discovery, `pypi:` add routing, and `run`/`exec` environment PATH.
- PEP 440/508 requirements, PubGrub backtracking, extras, markers, dependency groups, interpreter/native-wheel compatibility, and standard `pylock.toml`.
- Shared throttled HTTP, URL-scoped auth, negotiated Simple JSON metadata, verified ZIP/CAS ingestion, store-index persistence, and download reporting.
- One artifact-ingestion lifecycle for npm tarballs, Cargo crates, Python wheels and runtime ZIPs. Both formats share cache validation, offline gating, projection and store-index publication, HTTP request setup, extraction retries, integrity checks and progress tracking; decoding stays format-specific.
- Cargo sparse-index requests now use the configured retry budget. Python metadata uses the shared network policy, including HTTP 408 and interrupted bodies, with credential isolation on every retried redirect.
- Real virtual environments with RECORD verification, internal/filename wheel-tag agreement, entry points and wheel data relocation, including CRLF script shebangs.
- Python index responses and offline cache reads are bounded. The compact `python-index-v2` cache requires one online metadata refresh after upgrading. Python requests do not inherit npm credentials; explicit `python.indexUrl` credentials remain URL-scoped and require HTTPS or loopback for both metadata and wheels.
- Locked wheel replay replenishes concurrent download slots within the configured network limit, preserves deterministic errors, and settles pending work before reporting failure. Empty inline Python manifest tables are supported, and broken environment links identify the missing path.
- Archive and network retry URLs are redacted in both v11 and v12. Rust network errors strip embedded request URLs, retry counters handle the maximum configured budget, and manual metadata redirects retain configured guards.
- A new `pnpm-install-coordinator` crate owns metadata transactions, concurrent preparation, settlement, publication and reverse rollback. Cargo and Python declare their metadata and produce prepared projections; `install` and mixed `add` use the same lifecycle. The crate has no ecosystem or resolver dependency.
- Cargo lockfiles/source configuration now publish only after all preparations succeed. Publication failures restore native metadata and prior Python links. Unpublished generations are cleaned up; failed rollback retains recovery resources.
- Frozen/offline replay and production/development projections from the complete lockfile.
- Managed storage paths are excluded from native workspace discovery. A repeated install with an in-workspace store cannot discover cached Cargo/Python manifests as new projects.

Validation: `just ready` (10,127 tests passed, 5 skipped), Dylint, and all pre-push checks (TypeScript build/lint, Rust formatting/Clippy/rustdoc/Dylint, spelling and TOML formatting). The v11 tarball suite passed all 40 tests and the v11 network suite passed all 46 tests. Python/network/archive regressions cover metadata limits, credential isolation, wheel tags/scripts, concurrent settlement, redirect guards, redaction and retry-counter overflow. Self-update tests pass after removing a duplicate retry-settings mapping.

Real PyPI imports included `requests[socks]` and native `orjson`; uv 0.12.10 independently installed the generated seven-package lockfile. The real mixed smoke project replays offline after the review fixes, imports Python packages and runs `cargo run --offline --locked`. The macOS/Windows path-diagnostic fixture failures were investigated. The test now uses canonical project paths and native separators for the junction target. macOS and Ubuntu passed on the preceding head; Windows is rerunning with the separator correction.

npm explicitly participates as an in-place installer. Its normal materialization and lifecycle-script behavior is unchanged; this is not a transaction for `node_modules` or a post-publication build scheduler. The ownership and extension contract is documented in `pnpm/plans/ECOSYSTEM_INSTALL.md`.

The implementation is registry-wheel based, like the current Cargo integration's bounded registry scope. Source builds, editable/local/git/URL requirements, universal locking, arbitrary third-party pylock imports, pip/keyring configuration, and recursive/filtered add remain unsupported. Publication is not a crash journal; older generations are retained.

All 11 npm-only benchmark scenarios were measured for both the Python integration and the install-coordinator refactor. The coordinator run used four warmups and 30 samples per target, with differences from -2.5% to +1.9%. The largest slowdown was rechecked with reversed order and 40 samples: 246.46 ms versus 246.85 ms (+0.2%). Larger slowdowns did not reproduce. Initial integration evidence is in `pnpm/plans/PYTHON_SPIKE.md`; coordinator evidence is in `pnpm/plans/ECOSYSTEM_INSTALL.md`.

The artifact refactor was measured separately against `0915dc0d14`, using Linux release binaries, the same pnpr binary, four warmups and 30 samples per target. All eleven scenarios completed successfully. These measurements do not establish a speedup.

| npm-only scenario | Baseline mean (ms) | Artifact refactor mean (ms) |
| --- | ---: | ---: |
| Fresh install, cold cache/store | 498.65 | 506.09 |
| Fresh install, hot cache/store | 249.59 | 251.34 |
| Fresh install, cold cache/hot store | 437.63 | 445.85 |
| Frozen restore, cold cache/store | 384.18 | 385.90 |
| Frozen restore, hot cache/store | 113.60 | 114.11 |
| Repeat install, hot cache/store | 5.92 | 5.91 |
| Repeat install, cold cache/hot store | 6.10 | 5.95 |
| Add, hot cache/store | 273.98 | 267.33 |
| Offline resolution, hot cache | 171.85 | 172.54 |
| GVS frozen restore, hot cache/store | 106.14 | 110.88 |
| Frozen restore, cold cache/store/pnpr | 430.34 | 433.50 |

Six scenarios were rechecked with reversed target order and 40 samples. The larger slowdowns did not reproduce. The warm fresh-install difference narrowed to 0.1%; cold-cache repeat installs changed direction between runs, with differences of about 0.1 ms.

| Rechecked scenario | Baseline mean (ms) | Artifact refactor mean (ms) |
| --- | ---: | ---: |
| GVS frozen restore, hot cache/store | 107.54 | 107.20 |
| Fresh install, cold cache/store | 516.50 | 507.89 |
| Fresh install, cold cache/hot store | 436.57 | 438.21 |
| Frozen restore, cold cache/store | 389.70 | 386.76 |
| Fresh install, hot cache/store | 253.50 | 253.82 |
| Repeat install, cold cache/hot store | 6.05 | 6.14 |

## Squash Commit Body

```text
Exercise the multi-ecosystem architecture through a usable Python
integration, not test-only metadata writers.

Keep Python requirement, marker, lockfile and environment semantics
separate from npm and Cargo. Reuse the install-wide HTTP/auth budget,
verified artifact ingestion, CAS and store index.

Extract a shared install lifecycle without ecosystem-specific branches.
Native tasks declare metadata footprints and return prepared projections.
Settle all work before publishing Cargo or Python state, and reverse
attempted publications before restoring metadata on failure.
Retain resources when rollback fails so recovery remains possible.

Enroll npm with its existing in-place materialization semantics, and
keep npm-only dispatch on its early path. Do not unify native resolvers,
lockfile formats, target identity or package layouts.

Consolidate archive ingestion below the ecosystem boundary. Share cache
validation, authenticated requests, extraction retries and publication
across tarballs and ZIPs, retaining tar streaming and ZIP decoding.
Test the shared contracts across both formats and preserve npm fast paths.

Use standard pylock.toml, independently validated with uv.
New features target pnpm v12 only. Shared archive URL-redaction fixes
also cover pnpm v11.

Related to pnpm/pnpm#14566 and pnpm/rfcs#34.
```

A local validation run exposed an intermittent staging approval-order assertion. Investigation is kept separate from this PR; no staging code changes are included. The release-note-only state passed all 10,127 tests on rerun.

## Checklist

- [x] I checked the referenced issue and verified that none of the PRs
  already linked to it solves it.
- [x] New features are implemented only in the Rust pnpm v12 CLI. Bug fixes
  are implemented in every affected version.
- [x] Added separate release notes for `pacquet` and `pnpm` with the affected v11 packages. Consolidated the unreleased Cargo installation, add, authentication and rollback notes into the existing Cargo feature entry.
- [x] Added or updated tests.
- [x] Updated the documentation.

---
Written by an agent (Codex, GPT-5).

## Summary by CodeRabbit

* **New Features**
  * Added opt-in Python package management through PyPI, including `pypi:` dependencies, `pyproject.toml`, `pylock.toml`, and managed virtual environments.
  * Added Cargo dependency support for `pnpm add`, including version-saving options.
  * Added mixed Node, Cargo, and Python install coordination with rollback protection.
* **Bug Fixes**
  * Improved workspace discovery by excluding stores, caches, virtual environments, and Python caches.
  * Added retry handling for ecosystem downloads and safer redirect validation.
  * Redacted credentials and signed URL parameters from retry logs and errors.
* **Documentation**
  * Added guidance for configuring and testing Python ecosystem support.

### CODE REVIEW

File: pnpm/crates/cli/src/cli_args/pipelines.rs [hunk 1]
```diff
@@ -439,31 +439,28 @@ impl InstallPipeline {
...
-                projects: crate::cargo_deps::CargoInstallProjects::Workspace(&workspace_inventory),
-                lockfile_policy: crate::cargo_deps::CargoLockfilePolicy::UseExisting,
-            },
+            http_client,
         );
-        ecosystem_install::EcosystemInstallCoordinator::new(node_install)
-            .with_install(cargo_install)
+        ecosystem_plan
+            .with_task(pnpm_install_coordinator::InstallTask::in_place(Vec::new(), node_install))
             .run()
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/cli/src/cli_args/pipelines.rs:464]: Declining metadata-only rollback for ordinary `install`. npm is explicitly enrolled with in-place semantics: `InstallTask::in_place` rolls back only its declared metadata, and this call deliberately declares none. The contract is documented in https://github.com/pnpm/pnpm/blob/e3b18f8e7d7346dfa4e89a49126a8bf87c4ed671/pnpm/plans/ECOSYSTEM_INSTALL.md#native-projections-and-limits.

The npm installer can already have changed `node_modules` and run lifecycle scripts. Restoring its current lockfile/modules metadata alone would not roll those effects back and can leave the metadata describing a different installed tree. Mixed `add` separately promises restoration of the requested manifest/lock edits and has an explicit footprint. A transactional npm materialization layer remains outside this PR; I am keeping the ordinary-install behavior explicit rather than broadening that promise.

---
Written by an agent (Codex, GPT-5).

File: pnpm/crates/cli/src/python.rs [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+                Err(error) => Err(error).into_diagnostic(),
+            },
+            None => Ok(()),
+        }
+    }
+
+    fn retain(self: Box<Self>) {
+        if let Some(environment) = self.environment {
+            let _ = environment.keep();
+        }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/cli/src/python.rs:298]: Confirmed as a retained-generation limitation, but deferring automatic reclamation from this PR. `retain` is called both after successful publication and after failed rollback; resources kept for recovery must not be deleted by that shared path. A Python process launched from an older generation can also import modules from it after a newer generation is published. An age or generation-count cutoff alone cannot establish that the old environment is unused.

The retention contract and the absence of automatic generation GC are documented in https://github.com/pnpm/pnpm/blob/e3b18f8e7d7346dfa4e89a49126a8bf87c4ed671/pnpm/plans/ECOSYSTEM_INSTALL.md. Safe reclamation needs an explicit pruning operation or a separate lifetime/lease policy. This is a deliberate scope deferral, not a claim that storage is currently bounded.

---
Written by an agent (Codex, GPT-5).

File: pnpm/crates/cli/src/python/registry.rs [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+                format!("Python index for {name} is not cached for offline resolution")
+            })?;
+            serde_json::from_slice::<CachedIndex>(&contents).into_diagnostic()?
+        } else {
+            let response = self
+                .client
+                .get_bytes_with_secure_auth_and_retry(
+                    index_url.as_str(),
+                    &self.auth,
+                    Some("application/vnd.pypi.simple.v1+json"),
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/cli/src/python/registry.rs]: Addressed in 7f9b74e926ac16338dac65ebc66081eb83fba8a8.

Bounded Python index responses at 64 MiB using the shared limited-body reader, including chunked/error responses. Offline cache reads are bounded too. The new cache stores raw JSON rather than a numeric byte array; its versioned namespace requires one online metadata refresh. Tests cover oversized network/cache input and compact offline reuse.

---
Written by an agent (Codex, GPT-5).

File: pnpm/crates/cli/src/python.rs [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+            roots.push((path.parent().expect("manifest has a parent").to_path_buf(), manifest));
+        }
+    }
+    if roots.is_empty() {
+        return Ok(Vec::new());
+    }
+    let interpreter: Interpreter =
+        host::run(&config.python.executable, "probe", serde_json::json!({})).await?;
+    let mut index: url::Url = config.python.index_url.parse().into_diagnostic()?;
+    let mut auth = (*config.auth_headers).clone();
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/cli/src/python.rs]: Addressed in 7f9b74e926ac16338dac65ebc66081eb83fba8a8.

Python now starts with an independent credential set. A repository-selected index cannot select user npm credentials for metadata or wheels; explicit `python.indexUrl` credentials remain URL-scoped. Regression tests inject a URL-scoped npm token and assert neither endpoint receives it, while explicit Basic authentication still succeeds.

---
Written by an agent (Codex, GPT-5).

File: pnpm/crates/cli/src/python/host.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    dist_info = roots.pop()
+    for name in files:
+        parts = PurePosixPath(name).parts
+        if not parts or any(part in (".", "..") for part in parts) or name.startswith("/") or "\\" in name:
+            raise ValueError("unsafe wheel path: " + name)
+    wheel = read_headers(files, dist_info + "/WHEEL")
+    if wheel["Wheel-Version"] != "1.0":
+        raise ValueError("unsupported Wheel-Version: " + str(wheel["Wheel-Version"]))
+    if wheel["Root-Is-Purelib"] not in ("true", "false"):
+        raise ValueError("invalid Root-Is-Purelib")
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/cli/src/python/host.py:57]: Addressed in 7f9b74e926ac16338dac65ebc66081eb83fba8a8.

Wheel inspection now requires internal `Tag` fields to equal the expanded filename tags. Tests reject missing, incompatible and extra tags with valid RECORD hashes, and accept compressed filename tags expanded into separate WHEEL fields.

---
Written by an agent (Codex, GPT-5).

File: pnpm/crates/cli/src/python.rs [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+    match fs::symlink_metadata(&link) {
+        Err(error) if error.kind() == io::ErrorKind::NotFound => Ok(None),
+        Err(error) => Err(error).into_diagnostic(),
+        Ok(_) => {
+            if !pnpm_fs::is_symlink_or_junction(&link).into_diagnostic()? {
+                bail!("pnpm will not replace an unmanaged Python environment: {}", link.display());
+            }
+            let target = root.join(pnpm_fs::read_symlink_dir(&link).into_diagnostic()?);
+            let target = dunce::canonicalize(target).into_diagnostic()?;
+            let managed = dunce::canonicalize(root.join(".pnpm/python-envs")).into_diagnostic()?;
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/cli/src/python.rs]: Addressed in 7f9b74e926ac16338dac65ebc66081eb83fba8a8.

Both canonicalization failures now identify the missing target/managed directory and the `.venv` link being validated. Tests cover dangling links and a missing managed directory. Unmanaged environments are still rejected, with no lockfile publication.

---
Written by an agent (Codex, GPT-5).

File: pnpm/crates/cli/src/python.rs [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+        fs::rename(&staged, root.join(".venv")).into_diagnostic()
+    }
+}
+
+fn percent_decode(value: &str) -> Result<String> {
+    url::form_urlencoded::parse(format!("value={}", value.replace('+', "%2B")).as_bytes())
+        .next()
+        .map(|(_, value)| value.into_owned())
+        .ok_or_else(|| miette::miette!("invalid Python index credential"))
+}
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/cli/src/python.rs]: Addressed in 7f9b74e926ac16338dac65ebc66081eb83fba8a8.

Removed the form-query decoder and reused the existing network URL percent decoder, extracting it from proxy handling without duplicating it. CLI tests cover literal ampersands/pluses, percent-encoded UTF-8, and a slashless index URL.

---
Written by an agent (Codex, GPT-5).

File: pnpm/crates/cli/src/python/host.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+            if parts[0].endswith(".data"):
+                if parts[0] != dist_info.removesuffix(".dist-info") + ".data" or len(parts) < 3 or parts[1] not in scheme:
+                    raise ValueError("invalid wheel data path: " + name)
+                destination = Path(scheme[parts[1]]).joinpath(*parts[2:])
+                executable = executable or parts[1] == "scripts"
+            else:
+                destination = site.joinpath(*parts)
+            contents = Path(source).read_bytes()
+            if executable and (contents.startswith(b"#!python\n") or contents.startswith(b"#!pythonw\n")):
+                contents = ("#!" + str(interpreter) + "\n").encode() + contents.split(b"\n", 1)[1]
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/cli/src/python/host.py]: Addressed in 7f9b74e926ac16338dac65ebc66081eb83fba8a8.

Placeholder shebangs now handle LF and CRLF for both `python` and `pythonw`, preserving the script body. The regression test verifies installed interpreter paths, unchanged non-placeholder scripts, and every rewritten RECORD digest/size. It fails on the previous binary's CRLF script.

---
Written by an agent (Codex, GPT-5).

File: pnpm/crates/cli/src/python/manifest.rs [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+            updated.insert_str(end, &format!("\n{key} = {array}\n"));
+        } else if original[span.clone()].trim_start().starts_with('{') {
+            updated.insert_str(span.end - 1, &format!(", {key} = {array}"));
+        } else {
+            bail!("cannot add {table_name}.{key} to this TOML table representation");
+        }
+    } else {
+        writeln!(updated, "\n[{table_name}]\n{key} = {array}")
+            .expect("writing to a String cannot fail");
+    }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/cli/src/python/manifest.rs:174]: Addressed in 7f9b74e926ac16338dac65ebc66081eb83fba8a8.

Inline-table insertion now adds a separator only when the parsed table has entries. End-to-end add tests cover empty/populated `project` and `dependency-groups` inline tables, then import the installed package. The empty-project case reproduces the previous TOML parse error.

---
Written by an agent (Codex, GPT-5).

File: pnpm/crates/tarball/src/archive_request.rs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    let client = if revision_addressed {
+        http_client.acquire_for_url_without_redirects_with_priority(package_url, priority).await
+    } else {
+        http_client.acquire_for_url_with_priority(package_url, priority).await
+    };
+    let mut request = client.get(package_url);
+    if let Some(value) = auth_header_for_package_download(auth_headers, package_url, package_id) {
+        request = request.header("authorization", value);
+    }
+    let sent = request.send().await;
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/tarball/src/archive_request.rs:30]: Addressed in 7f9b74e926ac16338dac65ebc66081eb83fba8a8.

The production pnpr path already constructs its client with `new_for_installs_with_redirect_guard` in `pnpr/crates/pnpr/src/resolver.rs`, using the same route context as the fetch allowlist. CLI clients intentionally do not impose a deployment allowlist. I kept that policy ownership and added an archive-boundary regression test: an allowlisted endpoint redirects to a disallowed target, the request fails, and the target receives zero requests. No second redirect implementation was added.

---
Written by an agent (Codex, GPT-5).

File: pnpm/crates/tarball/src/archive_request.rs [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    Reporter::emit(&LogEvent::FetchingProgress(FetchingProgressLog {
+        level: LogLevel::Debug,
+        message: FetchingProgressMessage::Started {
+            attempt: attempt + 1,
+            package_id: package_id.to_owned(),
+            size,
+        },
+    }));
+    let response = sent.map_err(|error| {
+        TarballError::FetchTarball(NetworkError { url: package_url.to_string(), error })
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/tarball/src/archive_request.rs]: Addressed in 7f9b74e926ac16338dac65ebc66081eb83fba8a8.

Shared archive retry tracing now uses redacted URLs and error Display, not raw Debug. Reporter retry URLs are redacted too. Network errors remove reqwest's embedded URL so walking the source chain cannot expose its credentials/query/fragment. The analogous v11 retry-log and tarball-size-error leaks are fixed as well. Tests verify redaction while the actual signed request URL remains unchanged; v11's 40-test tarball suite passes.

---
Written by an agent (Codex, GPT-5).

File: pnpm/crates/tarball/src/archive_retry.rs [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+            Ok(value) => {
+                emit_progress_fetched::<Reporter>(package_id, requester, progress_key);
+                return Ok(value);
+            }
+            Err(err) if !is_transient_error(&err) => return Err(err),
+            Err(err) if attempt >= max_retries => {
+                tracing::warn!(
+                    target: "pacquet::download",
+                    ?package_url,
+                    attempts = attempt + 1,
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/tarball/src/archive_retry.rs]: Addressed in 7f9b74e926ac16338dac65ebc66081eb83fba8a8.

Tracing counters widen to u64; public u32 attempt counters saturate. The shared network retry log had the same max-budget overflow and is fixed too. Tests run a failure then success with `u32::MAX` retries, and exercise archive request reporting at attempt `u32::MAX` without iterating billions of times.

---
Written by an agent (Codex, GPT-5).

**Comment by zkochan** [comment]
Review follow-up in 7f9b74e926ac16338dac65ebc66081eb83fba8a8:

- Addressed the inline threads with regression tests and replies. The existing npm in-place contract and retention of older Python environments remain deliberate limits, as explained in those threads.
- Addressed the summary-only findings: bounded concurrent locked-wheel downloads with settlement, user-facing changeset wording, clearer Python save-flag and lockfile-conflict errors, rendered lockfile comparison output, the Python fast-path rustdoc gate, and reuse of `Config::retry_opts()` in self-update verification.
- Slashless Python index URLs were already normalized; a credentialed-index regression test now pins that behavior. Cargo metadata rollback already belongs to the shared coordinator, so no duplicate adapter transaction was added. Optional redirect-body draining is deferred without evidence that it improves this workload.

Validation: 10,122 Rust tests passed (5 skipped), 40 v11 tarball tests passed, all pre-push checks passed. The real mixed project still replays offline with Python imports and Cargo execution working. Watching the updated review/CI run.

---
Written by an agent (Codex, GPT-5).

File: pnpm/crates/cli/src/python.rs [hunk 13]
```diff
@@ -0,0 +1,385 @@
...
+        }
+    }
+    if roots.is_empty() {
+        return Ok(Vec::new());
+    }
+    let interpreter: Interpreter =
+        host::run(&config.python.executable, "probe", serde_json::json!({})).await?;
+    let mut index: url::Url = config.python.index_url.parse().into_diagnostic()?;
+    // A repository-selected Python index must not select user-level npm credentials.
+    let mut auth = pnpm_network::AuthHeaders::default();
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/cli/src/python.rs]: Fixed in 756141829d92816a81768d4026e787c77f93291a.

Python's isolated credential map now opts into `AuthHeaders::with_secure_transport()`. Every lookup, including shared archive lookups and clones, requires TLS or loopback. This keeps policy with the credentials rather than adding a Python package-ID branch to archive ingestion. A request-level regression uses an HTTPS-scoped credential with an HTTP wheel URL at the same registry path and confirms no Authorization header is transmitted; lookup tests also cover HTTPS and loopback.

---
Written by an agent (Codex, GPT-5).

File: pnpm/crates/network/src/lib.rs [hunk 14]
```diff
@@ -628,38 +628,95 @@ impl ThrottledClient {
...
+            let target = response
+                .headers()
+                .get(reqwest::header::LOCATION)
+                .and_then(|location| location.to_str().ok())
+                .and_then(|location| response.url().join(location).ok());
+            if is_redirect_status(response.status())
+                && redirect_count < MAX_REDIRECT_HOPS
+                && let Some(target) = target
+            {
+                current_url = target.to_string();
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/network/src/lib.rs:710]: Fixed in 756141829d92816a81768d4026e787c77f93291a.

The configured redirect guard is now installed on both paired clients. The shared policy validates the target, then either follows it or returns the redirect response for manual handling. This preserves the guard without adding another redirect-policy implementation or changing the secure metadata helper's error type. Regression tests cover accepted/rejected manual redirects, zero requests to a denied target, and no retry of a guard rejection.

---
Written by an agent (Codex, GPT-5).

File: pnpm11/fetching/tarball-fetcher/src/remoteTarballFetcher.ts [hunk 15]
```diff
@@ -119,7 +119,7 @@ export function createDownloader (
             maxRetries: downloadRetryOpts.retries,
             method: 'GET',
             timeout,
-            url,
+            url: redactUrlForDisplay(url),
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm11/fetching/tarball-fetcher/src/remoteTarballFetcher.ts:122]: Fixed in 756141829d92816a81768d4026e787c77f93291a.

Network retry logs now redact the URL field and request URLs echoed in transport-error messages. Tests inspect the complete retry-log payload for signed URLs and credential-bearing URLs rejected by the transport. All 46 v11 network tests pass. The Rust retry URL fields are redacted too, and exhausted secure metadata requests remove reqwest's embedded URL.

---
Written by an agent (Codex, GPT-5).

**Comment by zkochan** [comment]
Follow-up review fixes are in 756141829d92816a81768d4026e787c77f93291a.

- Secure Python wheel credentials, guarded manual redirects, metadata error redaction, and v11 network retry-log redaction are covered in the inline replies.
- Locked-wheel buffering now replenishes completed slots without waiting for earlier downloads. The three-wheel/two-slot test blocks the first wheel until the third starts, and still verifies settlement on failure. Errors remain deterministic in lockfile order.
- The npm resolver now imports the shared percent decoder. Python add reuses install's static-metadata validation before changing the manifest. Release-note wording and lockfile diagnostic output are updated.
- Investigated both cross-platform failures on the previous head. The environment-path test now creates its targets and expected paths from the canonical project root, matching CLI discovery on macOS and Windows. The replacement jobs will verify that fix.

Two optional suggestions were not applied: Python lockfile restoration already belongs to the shared coordinator's declared metadata footprint, so adding adapter snapshots would duplicate ownership. The wheel candidate lookup is an internal invariant: locked packages are inserted by `Lockfile::seed` before fetching, and resolver-selected wheels come from that same candidate map. Existing tampered-lockfile tests reject invalid dependency graphs without weakening that invariant.

Validation: `just ready` passed 10,127 tests (5 skipped), 363 targeted Rust tests passed, and all 46 v11 network tests passed. All pre-push checks passed. Continuing to watch reviews and CI.

---
Written by an agent (Codex, GPT-5).

File: pnpm/crates/cli/src/python.rs [hunk 16]
```diff
@@ -0,0 +1,385 @@
...
+        }
+    }
+    if roots.is_empty() {
+        return Ok(Vec::new());
+    }
+    let interpreter: Interpreter =
+        host::run(&config.python.executable, "probe", serde_json::json!({})).await?;
+    let mut index: url::Url = config.python.index_url.parse().into_diagnostic()?;
+    // A repository-selected Python index must not select user-level npm credentials.
+    let mut auth = pnpm_network::AuthHeaders::default().with_secure_transport();
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/cli/src/python.rs:144]: Checked this against the locked reqwest 0.13.4 implementation. This redirect does not retain Authorization: `remove_sensitive_headers` compares scheme as well as hostname and effective port, and removes the header on any change. `TowerRedirectPolicy::on_request` invokes it before each redirected request. That includes HTTPS → HTTP on the same host and port.

Source: https://github.com/seanmonstar/reqwest/blob/v0.13.4/src/redirect.rs#L239-L255 and the `on_request` method in that file. The initial URL is separately gated by the transport-restricted auth map in 756141829d92816a81768d4026e787c77f93291a. No additional redirect loop is needed for the reported credential leak.

---
Written by an agent (Codex, GPT-5).
